xbe-loader/xiso_driver.c

```diff
@@ -0,0 +1,304 @@
+  int r2 = XReadFile(iso_handle, &destination[offset], IrpSp->Parameters.Read.Length, &numberOfBytesRead);
+
+  Irp->IoStatus.Information = numberOfBytesRead;
+#endif
+
+
+  Irp->IoStatus.Status = STATUS_SUCCESS;
+  IoCompleteRequest(Irp, IO_NO_INCREMENT);
+  debug_print("Returning from read\n");
+  return STATUS_SUCCESS;
+
+
+#endif
+
+}
+
+static __attribute__((__stdcall__)) NTSTATUS irp_control(IN PDEVICE_OBJECT DeviceObject, IN PIRP Irp) {
+  debug_print("irp_control IRQL: %d\n", KeGetCurrentIrql());
+
+  assert(KeGetCurrentIrql() == DISPATCH_LEVEL);
+
+  NTSTATUS status;
+  PIO_STACK_LOCATION IrpSp;
+
+  IrpSp = IoGetCurrentIrpStackLocation(Irp);
+
+  switch (IrpSp->Parameters.DeviceIoControl.IoControlCode) {
+  //FIXME: Handle these
+  case IOCTL_CDROM_GET_DRIVE_GEOMETRY:
+
+    DISK_GEOMETRY* g = Irp->UserBuffer;
+    unsigned int filesize;
+    XGetFileSize(iso_handle, &filesize);
+
+    //FIXME: Support multi-part ISOs
+
+    g->Cylinders.QuadPart = filesize / sector_size;
+    g->MediaType = RemovableMedia;
+    g->TracksPerCylinder = 1;
+    g->SectorsPerTrack = 1;
+    g->BytesPerSector = sector_size;
+
+    status = STATUS_SUCCESS;
+    break;
+  case IOCTL_CDROM_XBOX_SECURITY:
+    status = STATUS_SUCCESS;
+    break;
+  default:
+    debug_print("Unhandled device control: 0x%x\n", IrpSp->Parameters.DeviceIoControl.IoControlCode);
+    status = STATUS_INVALID_DEVICE_REQUEST;
+    break;
+  }
+
+  if (status != STATUS_PENDING) {
+    Irp->IoStatus.Status = status;
+    IoCompleteRequest(Irp, IO_NO_INCREMENT);
+  }
+
+  return status;
+}
+
+// FIXME: The following code is for asynchronous I/O?
+
+static void finish_read(PIRP Irp) {
+  Irp->IoStatus.Status = STATUS_SUCCESS;
+  IoCompleteRequest(Irp, IO_CD_ROM_INCREMENT);
+}
+
+static void start_read(PIRP Irp) {
+  PIO_STACK_LOCATION IrpSp = IoGetCurrentIrpStackLocation(Irp);
+
+  bool is_scatter_gather = Irp->Flags & IRP_SCATTER_GATHER_OPERATION;
+  bool is_cache_request = IrpSp->Flags & FSC_REQUEST;
+  if (!is_scatter_gather || is_cache_request) {
+    //FIXME: Handle a buffer copy?
+  } else {
+    //FIXME: Handle scatter gather?
+  }
+
+  finish_read(Irp);
+}
+
+static __attribute__((__stdcall__)) VOID start_io(IN PDEVICE_OBJECT DeviceObject, IN PIRP Irp) {
+  assert(KeGetCurrentIrql() == DISPATCH_LEVEL);
+
+  debug_print("\nstart_io: Yay!\n\n");
+  Irp->IoStatus.Status = STATUS_REQUEST_ABORTED;
+  IoCompleteRequest(Irp, IO_NO_INCREMENT);
+
+  // Just avoid handling requests during shutdown
+  if (HalIsResetOrShutdownPending()) {
+    Irp->IoStatus.Status = STATUS_REQUEST_ABORTED;
+    IoCompleteRequest(Irp, IO_NO_INCREMENT);
+    return;
+  }
+
+  PIO_STACK_LOCATION IrpSp = IoGetCurrentIrpStackLocation(Irp);
+
+  switch (IrpSp->MajorFunction) {
+
+  case IRP_MJ_READ:
+    start_read(Irp);
+    break;
+  case IRP_MJ_DEVICE_CONTROL:
+    switch (IrpSp->Parameters.DeviceIoControl.IoControlCode) {
+    //FIXME: Handle these
+    default:
+      assert(false)
```

Comment:
Missing `;`?

Reply:
jup. I assume the code still works because I define asserts away. Will fix in the future